src/components/forms/cbr/CbrRuleForm.js

```diff
@@ -0,0 +1,292 @@
+import { TextArea } from "@carbon/react";
+import { transpose } from "lazy-z";
+import PropTypes from "prop-types";
+import React, { Component } from "react";
+import { IcseSelect } from "../../Dropdowns";
+import IcseFormTemplate from "../../IcseFormTemplate";
+import { IcseNameInput, IcseTextInput } from "../../Inputs";
+import { IcseFormGroup } from "../../Utils";
+import {
+  buildFormDefaultInputMethods,
+  buildFormFunctions,
+} from "../../component-utils";
+import CbrContextForm from "./CbrContextForm";
+import CbrResourceAttributeForm from "./CbrResourceAttributeForm";
+import CbrTagForm from "./CbrTagForm";
+
+/**
+ * Context-based restriction rules
+ */
+class CbrRuleForm extends Component {
+  constructor(props) {
+    super(props);
+    this.state = { ...this.props.data };
+    this.handleInputChange = this.handleInputChange.bind(this);
+    buildFormDefaultInputMethods(this);
+    buildFormFunctions(this);
+  }
+
+  handleInputChange(event) {
+    let { name, value } = event.target;
+    this.setState({ [name]: value });
+  }
+
+  render() {
+    // set up props for subforms
+    let contextInnerFormProps = {
+      invalidCallback: this.props.invalidContextCallback,
+      invalidTextCallback: this.props.invalidContextTextCallback,
+      arrayParentName: this.props.data.name,
+    };
+    transpose({ ...this.props.contextProps }, contextInnerFormProps);
+
+    let resourceAttributeInnerFormProps = {
+      invalidCallback: this.props.invalidResourceAttributeCallback,
+      invalidTextCallback: this.props.invalidResourceAttributeTextCallback,
+      arrayParentName: this.props.data.name,
+    };
+    transpose(
+      { ...this.props.resourceAttributeProps },
+      resourceAttributeInnerFormProps
+    );
+
+    let tagInnerFormProps = {
+      invalidCallback: this.props.invalidTagCallback,
+      invalidTextCallback: this.props.invalidTagTextCallback,
+      arrayParentName: this.props.data.name,
+    };
+    transpose({ ...this.props.tagProps }, tagInnerFormProps);
+
+    return (
+      <div id="cbr-rule-form">
+        <IcseFormGroup>
+          <IcseNameInput
+            id={this.props.data.name + "-cbr-rule"}
+            componentName={this.props.data.name + "-cbr-rule"}
+            value={this.state.name}
+            onChange={this.handleInputChange}
+            hideHelperText={true}
+            invalid={this.props.invalidCallback("name", this.state, this.props)}
```

Comment:
let's have a separate callback for name, `invalidNameCallback` and  `invalidNameTextCallback` that accepts only `this.state` and `this.props` so we can have it be consistant with craig name callback and text callback function

Reply:
should i do this for the subforms as well?